From d544f5674e223b255b52d0ffed1856e042c1ce2a Mon Sep 17 00:00:00 2001
From: Wendy Liang <wendy.liang@xilinx.com>
Date: Thu, 24 Sep 2020 11:36:01 -0700
Subject: [PATCH 1660/1852] misc: xilinx-ai-engine: Add error category to the
 AIE error struct

commit 36ed8979779ddc1cf7d74e053aac6aed38890ca9 from
https://github.com/Xilinx/linux-xlnx.git

Add error category property to the AI engine error struct so that
caller can get the category of a specific AI engine error event.

Signed-off-by: Wendy Liang <wendy.liang@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../xilinx-ai-engine/ai-engine-interrupt.c    |  8 +++++--
 include/linux/xlnx-ai-engine.h                | 23 +++++++++++++++++++
 2 files changed, 29 insertions(+), 2 deletions(-)

diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c b/drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c
index 77985118f014..cc7b9f925884 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-interrupt.c
@@ -763,8 +763,11 @@ static u32 aie_get_errors_from_bitmap(struct aie_partition *apart,
 	u32 num_err = 0;
 
 	for (i = 0; i < err_attr->num_err_categories; i++) {
-		for (j = 0; j < err_attr->err_category[i].num_events; j++) {
-			u8 event = err_attr->err_category[i].prop[j].event;
+		const struct aie_err_category *category;
+
+		category = &err_attr->err_category[i];
+		for (j = 0; j < category->num_events; j++) {
+			u8 event = category->prop[j].event;
 
 			if (!aie_check_error_bitmap(apart, loc, module, event))
 				continue;
@@ -773,6 +776,7 @@ static u32 aie_get_errors_from_bitmap(struct aie_partition *apart,
 			aie_err[num_err].loc.row = loc.row;
 			aie_err[num_err].module = module;
 			aie_err[num_err].error_id = event;
+			aie_err[num_err].category = category->err_category;
 			num_err++;
 		}
 	}
diff --git a/include/linux/xlnx-ai-engine.h b/include/linux/xlnx-ai-engine.h
index a030d578b8c3..9f489305238e 100644
--- a/include/linux/xlnx-ai-engine.h
+++ b/include/linux/xlnx-ai-engine.h
@@ -66,10 +66,18 @@ enum aie_module_type {
 
 struct device;
 
+/**
+ * struct aie_error - AI engine error
+ * @loc: AI engine tile location of which the error is from
+ * @module: AI engine module type of which the error is from
+ * @error_id: AI engine hardware event ID
+ * @category: AI engine error category of the error
+ */
 struct aie_error {
 	struct aie_location loc;
 	enum aie_module_type module;
 	u32 error_id;
+	u32 category;
 };
 
 struct aie_errors {
@@ -95,6 +103,17 @@ const char *aie_get_error_string(struct aie_errors *aie_errs,
 				 struct aie_error *aie_err);
 int aie_flush_errors(struct device *dev);
 void aie_free_errors(struct aie_errors *aie_errs);
+
+/**
+ * aie_get_error_category() - Get the category of an AIE error
+ * @err: AI engine hardware error
+ * @return: category of the error
+ */
+static inline u32 aie_get_error_category(struct aie_error *err)
+{
+	return err->category;
+}
+
 #else
 static inline bool aie_partition_is_available(struct aie_partition_req *req)
 {
@@ -159,5 +178,9 @@ static inline int aie_flush_errors(struct device *dev)
 
 static inline void aie_free_errors(struct aie_errors *aie_errs) {}
 
+static inline u32 aie_get_error_category(struct aie_error *err)
+{
+	return 0;
+}
 #endif /* CONFIG_XILINX_AIE */
 #endif
-- 
2.31.1

